### PR TITLE
expose extraQueryParameters

### PR DESCRIPTION
**Pull request by siddhijain**

### CODE REVIEW

**Review by Avery-Dunn — APPROVED** [review]

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/TokenRequestExecutor.java [hunk 1]
```diff
@@ -67,6 +67,15 @@ OAuthHttpRequest createOauthHttpRequest() throws SerializeException, MalformedUR
             params.put("claims", Collections.singletonList(claimsRequest));
         }
 
+        if(msalRequest.requestContext().apiParameters().extraQueryParameters() != null ){
+            for(String key: msalRequest.requestContext().apiParameters().extraQueryParameters().keySet()){
+                    if(params.containsKey(key)){
```
Comment (bgavrilMS) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/TokenRequestExecutor.java:72]: There is no need to add this validation. If app developer sets this and overrides some params that MSAL computes, it's their problem. 

Consider adding just a `log.warning`

File: msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java [hunk 2]
```diff
@@ -178,9 +175,12 @@ public void acquireTokenSilent_ConfidentialClient_acquireTokenSilent(String envi
         cfg = new Config(environment);
 
         IConfidentialClientApplication cca = getConfidentialClientApplications();
-
+        //test that adding extra query parameters does not break the flow
+        Map<String, String> extraParameters = new HashMap<>();
+        extraParameters.put("test","test");
         IAuthenticationResult result = cca.acquireToken(ClientCredentialParameters
                 .builder(Collections.singleton(cfg.graphDefaultScope()))
+                        .extraQueryParameters(extraParameters)
```
Comment (bgavrilMS) [inline msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java:183]: Is there a way to assert this? Maybe by intercepting the HTTP request message and asserting that the URL indeed contains `test=test` ?

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/TokenRequestExecutor.java (hunk 1 above)
Comment (siddhijain) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/TokenRequestExecutor.java:72]: Sure I was going by the requirements, thinking this is what all the MSALs do. https://github.com/AzureAD/microsoft-authentication-library-for-java/issues/256 Point 5.

File: msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java (hunk 2 above)
Comment (siddhijain) [reply msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java:183]: I wanted to do that too - One way of intercepting these calls is by mocking HTTP requests (which we do not prefer). These tests do not mock the HTTP requests and hence, the request cannot be intercepted. I will check to see if I can find another way.

Comment (bgavrilMS) [reply msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java:183]: I believe all MSAL libraries allow injection of custom HTTP interface - https://github.com/AzureAD/microsoft-authentication-library-for-java/wiki/Configure-Http-Client and this seems like a good way to write unit tests ?

Comment (siddhijain) [reply msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java:183]: That's a good point. By providing a simple CustomHttpClient, I can **build** an app that asserts that the httpRequest in the httpClient contains the extraQueryParameters.  

However, if we need to test end-to-end (and actually acquire tokens using httpClient), providing a custom HTTP client needs quite some code to be written to make get and post requests and process responses. https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/dev/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/DefaultHttpClient.java

Is it worth writing this much code to test this? Thoughts?

Comment (bgavrilMS) [reply msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java:183]: I was thinking just mocking IHttpClient like this - https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/00c991a91f154cb7bad379da21504444c2405ee6/msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/ConfidentialClientApplicationUnitT.java#L246

Not suggesting you to have an E2E test here, unit tests only. 

Most (>95%) of MSAL.NET's unit tests are written like this: 

- use a mock IHttpClient (see [MockHttpManager ](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/03fd41da88d9c23d15dd800ff57f63d2a2ffecea/tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpManager.cs#L21))
- the mock object has a queue of (request, response) objects which the tests sets
- on each `send`, you dequeue a (request, response) and serve it from there. 
- also have an option to assert the request, so for example you can see the actual URI 

There may be a lot of code at first, but it's good infrastructure that you can reuse later.

Comment (siddhijain) [reply msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java:183]: Got it. Thanks for clarifying. I was reluctant to go down the route of mocking httpClient using EasyMock, as later in those tests, we use PowerMock to assert results. And we are planning to stop using PowerMock in the test cases. Let me explore using the JUnit framework to write this test case and achieve the same result.

Comment (siddhijain) [reply msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java:183]: Updated an existing test case to capture the HTTP request and validate the parameters. https://github.com/AzureAD/microsoft-authentication-library-for-java/pull/607/files#:~:text=Assert.assertTrue(body.contains(%22test%3Dtest%22))

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/TokenRequestExecutor.java (hunk 1 above)
Comment (siddhijain) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/TokenRequestExecutor.java:72]: Updated in the latest commit.

File: msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java (hunk 2 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/AcquireTokenSilentIT.java:183]: LGTM. Sometimes these mocking frameworks are too complex for their own good. At the end of the day, it's just about adding a test implementation of IHttpClient

**Review by bgavrilMS — APPROVED** [review]

**Pull request merged**